src/coreclr/jit/flowgraph.cpp

```diff
@@ -5998,38 +6062,193 @@ void FlowGraphNaturalLoop::Duplicate(BasicBlock** insertAfter, BlockToBlockMap*
+                }
+
+                region = comp->ehGetEnclosingRegionIndex(region, &inTry);
+
+                if (region == EHblkDsc::NO_ENCLOSING_INDEX)
+                {
+                    break;
+                }
+            }
+        }
+    }
 
-    BasicBlock* bottom = GetLexicallyBottomMostBlock();
+    // Keep track of how much the region end EH indices change because of EH region cloning.
+    //
+    unsigned ehRegionShift = 0;
+
+    // Keep track of which blocks were handled by EH region cloning
+    //
+    BitVecTraits traits(comp->compBasicBlockID, comp);
+    BitVec       visited(BitVecOps::MakeEmpty(&traits));
+
+    VisitLoopBlocksLexical([=, &traits, &visited, &clonedTry, &ehRegionShift](BasicBlock* blk) {
+        if (canCloneTry)
+        {
+            // If we allow cloning loops with EH, we may have already handled
+            // this loop block as part of a containing try region.
+            //
+            if (BitVecOps::IsMember(&traits, visited, blk->bbNum))
+            {
+                return BasicBlockVisit::Continue;
+            }
+
+            // If this is a try region entry, clone the entire region now.
+            // Defer adding edges and extending EH regions until later.
+            //
+            // Updates map, visited, and insertAfter.
+            //
+            if (comp->bbIsTryBeg(blk))
+            {
+                Compiler::CloneTryInfo info(traits, visited);
+                info.m_map          = map;
+                info.m_addEdges     = false;
+                info.m_profileScale = weightScale;
+
+                BasicBlock* const clonedBlock = comp->fgCloneTryRegion(blk, info, insertAfter);
+
+                assert(clonedBlock != nullptr);
+                ehRegionShift += info.m_ehRegionShift;
+                clonedTry = true;
+                return BasicBlockVisit::Continue;
+            }
+        }
+        else
+        {
+            // We're not expecting to find enclosed EH regions
+            //
+            assert(!comp->bbIsTryBeg(blk));
+            assert(!comp->bbIsHandlerBeg(blk));
+            assert(!BitVecOps::IsMember(&traits, visited, blk->bbNum));
+        }
 
-    VisitLoopBlocksLexical([=](BasicBlock* blk) {
-        // Initialize newBlk as BBJ_ALWAYS without jump target, and fix up jump target later
-        // with BasicBlock::CopyTarget().
-        BasicBlock* newBlk = comp->fgNewBBafter(BBJ_ALWAYS, *insertAfter, /*extendRegion*/ true);
+        // `blk` was not in loop-enclosed try region or companion region.
+        //
+        // Initialize newBlk as BBJ_ALWAYS without jump target; these are fixed up subsequently.
+        //
+        // CloneBlockState puts newBlk in the proper EH region. We will fix enclosing region extents
+        // once cloning is done.
+        //
+        BasicBlock* newBlk = comp->fgNewBBafter(BBJ_ALWAYS, *insertAfter, /* extendRegion */ false);
```

Comment:
Seems very complicated. Is it possible to extract this to a separate function so that the actual enabled-by-default logic remains readable?

Reply:
Probably so, though I'd like to have the second use case far enough along that whatever I do here works nicely for that case too.

Reply:
I'm mostly worried about the maintainability. I spent a fair amount of effort untangling this logic from what it used to be in loop cloning.